File: 2015/day5/deno/logic.ts
/**
 * Checks if there are three or more vowels in a string.
 * @param {string} datum String to check.
 */
function hasThreeOrMoreVowels(datum: string): boolean {
    return (datum.match(/([aeiou])/g) || []).length > 2;
}

/**
 * Checks if there is a pair of letters repeated in a string.
 * @param {string} datum String to check.
 */
function hasSameLetterPair(datum: string): boolean {
    return (datum.match(/(.)\1{1,}/g) || []).length > 0;
}

/**
 * Checks if there is any naughty strings in the string.
 * @param {string} datum String to check.
 */
function hasNoNaughtyStrings(datum: string): boolean {
    return (datum.match(/(ab)|(cd)|(pq)|(xy)/g) || []).length < 1;
}

/**
 * Checks if string is nice.
 * @param {string} datum String to check.
 */
function isStringNice(datum: string): boolean {
    return hasThreeOrMoreVowels(datum) &&
        hasSameLetterPair(datum) &&
        hasNoNaughtyStrings(datum);
}

export { hasThreeOrMoreVowels, hasSameLetterPair, hasNoNaughtyStrings, isStringNice }

File: 2015/day5/ruby/logic.rb
##
# Checks if there are three or more vowels in a string.
def has_three_or_more_vowels(datum)
  datum.scan(/([aeiou])/).length > 2
end

##
# Checks if there is a pair of letters repeated in a string.
def has_same_letter_pair(datum)
  /(.)\1{1,}/.match?(datum)
end

##
# Checks if there is any naughty strings in the string.
def has_no_naughty_strings(datum)
  /(ab)|(cd)|(pq)|(xy)/.match?(datum) == false
end

##
# Checks if string is nice.
def is_string_nice(datum)
  has_three_or_more_vowels(datum) and
    has_same_letter_pair(datum) and
    has_no_naughty_strings(datum)
end
